policy: Support CIDRs in rules with zero length prefix

joestringer opened the pull request:
There's no reason to disallow specifying 0.0.0.0/0 CIDRs in policies,           
they're as valid as any other CIDR. Make the /0 check more specific to          
deny non-contiguous masks, which is what we actually disallow.                  
                                                                                
Fixes: #3146                                                                    
Fixes: #3945

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4458)
<!-- Reviewable:end -->

joestringer committed:
policy: Make CIDRRule error more consistent

Provide more information when the parsed Cidr is in an invalid format.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Support CIDRs in rules with zero length prefix

There's no reason to disallow specifying 0.0.0.0/0 CIDRs in policies,
they're as valid as any other CIDR. Make the /0 check more specific to
deny non-contiguous masks, which is what we actually disallow.

Fixes: #3146
Fixes: #3945

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf reviewed (dismissed):
Joe is doing additional testing

joestringer committed:
labels: Fix source for existing cidr tests

These tests assumed that labels are specified by source:key=value, but
for label arrays they're actually handled as source.key=value. Fix up
the test. This requires no functional changes.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
labels: Resolve CIDR 0.0.0.0/0 to reserved:world

Previously, we would resolve 0.0.0.0/0 (or ::/0) to a set of labels
0.0.0.0/0, reserved:world. The policy handling for this would be
expected to be exactly the same as for reserved:world, however
previously this would lead to allocating a unique CIDR identity for this
combination of labels. To prevent an identity conflict between this and
resolved:world, drop the 0.0.0.0/0 (or ::/0) label during CIDR to labels
conversion.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
ipcache: Don't push reserved identities to kvstore

Reserved identities such as the one for world shouldn't be pushed to the
kvstore, as they will always be handled locally as the daemon inserts
such IP->identity mappings into the local ipcache on startup.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Allow 0/0 CIDR to match reserved:world

When generating the EndpointSelector for a CIDRRule, if the CIDR selects
0.0.0.0/0 or ::/0, then add also the selector for reserved:world as
these should be equivalent.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
test: Add runtime policy test for 0.0.0.0/0

Add a runtime test to check that policies allowing 0.0.0.0/0 actually
allow connectivity to the outside world.

Signed-off-by: Joe Stringer <joe@covalent.io>
